refactor(api): ot3 hardware control API

sfoster1 opened the pull request:
Switches the OT3 hardware controller to a whole new hardware control API rather than just a controller.

Some changes required for the OT3 to work - the big one is that the plunger runs in the other direction - really can't be entirely handled in the controller layer alone, and need hardware API changes. With #9201 , it's much more reasonable to have two hardware control APIs with the same interface and working with typechecking, so this PR creates that second API and adds the changes required for the OT3 to run protocols. 

It also adds testing for the new API, by extending the `hardware` fixture in the API tests to provide both ot3 and ot2 hardware controllers. To me, this testing is sufficient to not have to do the same thing in the robot server, but I'm interested in other opinions.

The next steps after this API is to start pulling common implementation into shared code rather than leaving it duplicated between the two APIs, and either merge the controller code into the API or (since we may need the Simulation layer until we can finally get rid of api-level simulation) change the interface to better-reflect the internal data structures of the OT3 hardware controller.

This PR when merged should allow protocols to be run on the OT3 as long as you're ok with not homing, having hardcoded pipettes, and changing the default deck calibration to [[1, 0, 0], [0, -1, 0], [0, 0, -1]] and the pipette calibration to have different offsets. Those defaults will follow in other PRs.

Cannot be merged until
- [x] https://github.com/Opentrons/opentrons/pull/9201 is merged
- [x] https://github.com/Opentrons/opentrons/pull/9207 is merged
- [x] https://github.com/Opentrons/opentrons/pull/9209 is merged
- ~[ ] targeted on edge instead of https://github.com/Opentrons/opentrons/tree/ot3-protocol-base which is a synthetic merge-base of edge + the above three prs~ retargeted on ot3-5.0-work
- ~[ ] 5.0.0 will no longer be including changes in edge by default~ retargeted on ot3-5.0-work

Commit:
feat(shared-data):  Add gen3 pipettes

- Update liquid handling functions for gen3
- Update dev type unions
- Update python bindings to include gen3

Support gen3 names in shared data.

The positions now use the 0 origin at the top, and go positive towards
the bottom.

sfoster1 committed:
refactor(api): use ot3 pipettes on ot3

When we're using the OT3 hardware control option, the controller will
hardcode attached gen3 pipettes. This also requires support in the
protocol engine pipette name enums for gen3 pipettes.

shlokamin reviewed (changes requested):
this is raaaad! we'll need to add these new pipettes to the block lists for pipette dropdowns in PD + the app so they don't show up. I can do that if you want!

See the `PipetteSelect` in shared components, it takes a `nameBlocklist` prop that consuming components in PD + the app should now be utilizing.

sfoster1 committed:
chore(build): Fix push-ot3

the push-ot3 targets weren't properly accomodating the ot3 firmware
wheel, and in general could stand to handle sdists a little better.

mcous commented on api/src/opentrons/protocol_engine/resources/deck_data_provider.py:
```diff
@@ -37,8 +37,7 @@ def __init__(self, labware_data: Optional[LabwareDataProvider] = None) -> None:
     @staticmethod
     async def get_deck_definition(short_fixed_trash: bool = False) -> DeckDefinitionV2:
```
This `short_fixed_trash` arg is entirely unused by the ProtocolEngine code. Since we're consolidating FF checking into `deck_type`, which is nice, we can safely remove this arg (and the two tests in `tests/opentrons/protocol_engine/resources/test_deck_data_provider.py` that use it can also be removed)

mcous commented on api/src/opentrons/protocols/geometry/deck.py:
```diff
@@ -42,7 +42,7 @@ class CalibrationPosition:
 
 class Deck(UserDict):
-    def __init__(self, load_name: str = STANDARD_DECK) -> None:
+    def __init__(self, load_name: str = "") -> None:
```
Can this be optional, instead of empty-string-as-falsey?

```python
    def __init__(self, load_name: Optional[str] = None) -> None:
```

mcous commented on api/tests/opentrons/protocol_engine/conftest.py:
```diff
@@ -6,14 +6,17 @@
-from opentrons.protocols.api_support.constants import STANDARD_DECK, SHORT_TRASH_DECK
+from opentrons.protocols.api_support.constants import (
+    STANDARD_OT2_DECK,
+    SHORT_TRASH_DECK,
```
Following up on the comment above about FF logic consolidation; this import (and the `short_trash_deck_def` below) can be removed if that unused arg is removed

sfoster1 replied:
This code moved to https://github.com/Opentrons/opentrons/pull/9207

mcous reviewed (approved):
PE stuff looks good, just a couple comments about things this PR makes redundant

sfoster1 committed:
refactor(ot3): fix home positions

The home positions and directions of motion were all a little screwed
up; now they are not, and z positive is properly down, and everything
gets converted right.

sfoster1 committed:
BREAKS EVERYTHING

do NOT merge this but you need it for ot3 aspirate. The aspirate needs
to go the other direction.

amitlissack commented on api/src/opentrons/hardware_control/hardware_protocols.py:
```diff
@@ -0,0 +1,655 @@
+        """
+        ...
+
+    def get_attached_instrument(self, mount: Mount) -> "PipetteDict":
```
Does `PipetteDict` need to be quoted?

sfoster1 replied:
I think not since we put type extensions on the robot

sfoster1 replied:
This code moved to https://github.com/Opentrons/opentrons/pull/9201

amitlissack commented on api/src/opentrons/hardware_control/hardware_protocols.py:
```diff
@@ -0,0 +1,655 @@
+
+    # Gantry/frame (i.e. not pipette) action API
+    async def home_z(self, mount: Mount = None) -> None:
+        """Home the two z-axes"""
```
If this homes both z axes why does it take a `mount` argument?

sfoster1 replied:
docstring's out of date

amitlissack commented on api/src/opentrons/hardware_control/hardware_protocols.py:
```diff
@@ -0,0 +1,655 @@
+        ...
+
+    async def home(self, axes: List[Axis] = None) -> None:
+        """Home the entire robot and initialize current position.
```
nit. this function homes a list of axes.

amitlissack commented on api/src/opentrons/hardware_control/hardware_protocols.py:
```diff
@@ -0,0 +1,655 @@
+        # position reporting when motors are not homed
+        fail_on_not_homed: bool = False,
+    ) -> Point:
+        ...
```
Doc string?

sfoster1 committed:
refactor(app,components,protocol-designer): pgen3

Support (by blocklisting) Gen3 pipettes in all frontends

sfoster1 committed:
refactor(hardware): add "motion control"

Adds super basic planning code to hardware. Also changes move output to
use one move group per call to hardware move() rather than one per
actual move.

sfoster1 committed:
feat(hardware): Add load testing scripts

These scripts, when run, will generate a lot of CANbus messages and test
the behavior of the bus under load.

sfoster1 committed:
refactor(api): add module polling to ot3

The module polling actually lives in the controller and hadn't
previously been supported in the ot3controller. Now it is.

sfoster1 commented:
> this is raaaad! we'll need to add these new pipettes to the block lists for pipette dropdowns in PD + the app so they don't show up. I can do that if you want!
> 
> See the `PipetteSelect` in shared components, it takes a `nameBlocklist` prop that consuming components in PD + the app should now be utilizing.

This code moved to https://github.com/Opentrons/opentrons/pull/9207

sfoster1 commented:
> this is raaaad! we'll need to add these new pipettes to the block lists for pipette dropdowns in PD + the app so they don't show up. I can do that if you want!
> 
> See the `PipetteSelect` in shared components, it takes a `nameBlocklist` prop that consuming components in PD + the app should now be utilizing.

This code moved to https://github.com/Opentrons/opentrons/pull/9209

sfoster1 committed:
add p20 to applicable lists

sfoster1 committed:
refactor(api): ot3 gets its own HC API

With a mypy protocol defining the hardware API and some behavioral
changes that require alterations above the level of the controller, it's
both a better and safe idea to make the OT3 use a different hardware
control object rather than just a different controller. This PR removes
the OT3-specific changes and use of the OT3 controller from
hardware_control.api and puts them in hardware_control.ot3api.

sfoster1 committed:
Merge branch 'refactor_add-gen3-pipettes' into ot3-protocol-base

sfoster1 committed:
refactor(api): Test ot3 controller

By adapting the hardware fixture to provide both ot2 and ot3 apis, we
get free regression testing! There's a couple functions that outright
won't work on the other one, and a mark is provided for that. A lot of
tests had to be fixed up to mock methods of the injected hardware object
rather than the class, and I took the time to switch them to use
mock.patch which I like way more than pytest monkeypatch fixture.

shlokamin reviewed (approved):
sounds like the pipette select work got pulled from this PR so we can worry about that elsewhere

Laura-Danielle commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -0,0 +1,1910 @@
+        """
+        return self._backend.get_lights()
+
+    async def identify(self, duration_s: int = 5):
```
Should we remove this since the LED stuff will be different?

sfoster1 replied:
We should maybe remove the specific LED stuff (though I think we should wait on that) but `identify` in and of itself is actually a kind of generic semantic concept that can stay.

Laura-Danielle commented on hardware/tests/test_scripts/test_can_comm.py:
```diff
@@ -101,18 +105,19 @@ def test_prompt_payload_bad_input(
     mock_get_input.side_effect = user_input
     with pytest.raises(can_comm.InvalidInput):
-        can_comm.prompt_payload(DeviceInfoResponsePayload, mock_get_input)
+        can_comm.prompt_payload(GetStatusResponsePayload, mock_get_input)
```
Why did we change this to `GetStatus`?

sfoster1 replied:
The test changed because the device info prompt isn't the same as it used to be. There are probably other ways to change it though.

Laura-Danielle reviewed (approved):
Had two comments, in general looks good.
